File: GFG POTD/020-July-2025/017-power-of-k-in-n-where-k-may-be-non-prime.cpp
// https://www.geeksforgeeks.org/problems/power-of-k-in-n-where-k-may-be-non-prime4206/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  private:
    // TC : O(sqrt(k))
    // SC : O(sqrt(k))
    unordered_map<int, int> getPrimeFactorizationK(int k) {
        unordered_map<int, int> mpp;
        for (int i = 2; i * i <= k; i ++) {
            while (k % i == 0) {
                mpp[i] ++;
                k /= i;
            }
        }
        if (k > 1) mpp[k] ++;
        return mpp;
    }
    
    // Legendre's Formula to calculate the no. of time prime 'P' occurs in 'N'
    int primePCountInN(int n, int p) {
        int count = 0;
        while (n) {
            n /= p;
            count += n;
        }
        return count;
    }
  public:
    // TC : O(sqrt(k) + m*log(n))
    // SC : O(sqrt(k))
    int maxKPower(int n, int k) {
        unordered_map<int, int> primeFactorK = getPrimeFactorizationK(k);
        
        int result = INT_MAX;
        for (auto &it : primeFactorK) {
            int primeFactor    = it.first;
            int primeFactorCnt = it.second;
            
            int occurances = primePCountInN(n, primeFactor);
            
            result = min(result, occurances / primeFactorCnt);
        }
        return result;
    }
};

File: GFG POTD/020-July-2025/017-power-of-k-in-n-where-k-may-be-non-prime.java
// https://www.geeksforgeeks.org/problems/power-of-k-in-n-where-k-may-be-non-prime4206/1




import java.util.*;

class Solution {
    // TC : O(sqrt(k) + m*log(n))
    // SC : O(sqrt(k))
    public int maxKPower(int n, int k) {
        Map<Integer, Integer> primeFactorsK = getAllPrimeFactor(k);
        int result = Integer.MAX_VALUE;
        for (Map.Entry<Integer, Integer> it : primeFactorsK.entrySet()) {
            int primeFactor    = it.getKey();
            int primeFactorCnt = it.getValue();
            
            int occurances = cntPrimePInN(n, primeFactor);

            result = Math.min(result, occurances / primeFactorCnt);
        }
        return result;
    }
    
    private int cntPrimePInN(int n, int p) {
        int count = 0;
        while (n > 0) {
            n /= p;
            count += n;
        }
        return count;
    }
    
    private Map<Integer, Integer> getAllPrimeFactor(int k) {
        Map<Integer, Integer> pf = new HashMap<>();
        for (int i = 2; i * i <= k; i ++) {
            while (k % i == 0) {
                pf.put(i, pf.getOrDefault(i, 0) + 1);
                k /= i;
            }
        }
        if (k > 1) pf.put(k, pf.getOrDefault(k, 0) + 1);
        return pf;
    }
}